src/groups/UpgradeableRenounceableProxy.sol

```diff
@@ -0,0 +1,63 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+pragma solidity >=0.8.24;
+
+import "@openzeppelin/contracts/proxy/ERC1967/ERC1967Proxy.sol";
+
+interface IUpgradeableRenounceableProxy {
+    function implementation() external view returns (address);
+    function upgradeToAndCall(address _newImplementation, bytes memory _data) external;
+    function renounceUpgradeability() external;
+}
+
+contract UpgradeableRenounceableProxy is ERC1967Proxy {
+    // Errors
+
+    error BlockReceive();
+
+    // Constants
+
+    /// @dev Initial proxy admin.
+    address internal immutable ADMIN_INIT;
+
+    // Constructor
+
+    constructor(address _implementation, bytes memory _data) ERC1967Proxy(_implementation, _data) {
+        // set the admin to the deployer
+        ERC1967Utils.changeAdmin(msg.sender);
+        // set the admin as immutable
+        ADMIN_INIT = msg.sender;
+    }
+
+    function implementation() external view returns (address) {
+        return _implementation();
+    }
+
+    /// @dev Dispatch if caller is admin.
+    function _fallback() internal virtual override {
+        if (msg.sender == ADMIN_INIT && msg.sender == ERC1967Utils.getAdmin()) {
+            _dispatchAdmin();
+        } else {
+            super._fallback();
+        }
+    }
```

Comment:
To ensure consistency, I suggest moving the implementation() view function call directly into the fallback function as well.
```suggestion

    /// @dev Handles proxy function calls: attempts to dispatch to a specific
    ///      function or delegates all calls to the implementation contract.
    function _fallback() internal virtual override {
        // staticcall implementation() returns the address
        if (msg.sig == IUpgradeableRenounceableProxy.implementation.selector) {
            bytes32 slot = ERC1967Utils.IMPLEMENTATION_SLOT;
            assembly {
                let implementation := sload(slot)
                mstore(0, shr(12, shl(12, implementation)))
                return(0, 0x20)
            }
        }
        // dispatch if caller is admin, otherwise delegate to the implementation
        if (msg.sender == ADMIN_INIT && msg.sender == ERC1967Utils.getAdmin()) {
            _dispatchAdmin();
        } else {
            super._fallback();
        }
    }
```

Reply:
so personally, over the weekend, I reflected on this and I do not think moving these function signatures in the fallback handler is really of much benefit. It complicates the code because the conflict still exists; it is just more difficult to reason about because it depends on who is calling it.

Unless you have an example of adoption as a pattern, in general I don't think this is a pattern we should adopt for the long term. Rather for this temporary use-case it is good to try, and I can see the benefit because the proxy is "renounceable", so then these hidden-fallback-functions indeed disappear effectively once renounced. So in that regard I like it, and in any case we can keep it for here.

Reply:
the main benefit is avoiding selector clashes: in case implementation contract has functions with the same selectors as we have inside the proxy - these functions are unreachable, as dispatcher finds a match inside the proxy and never gets to the fallback to redirect call to the implementation.

openzeppelin provides an example of adoption as a pattern: https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/proxy/transparent/TransparentUpgradeableProxy.sol#L95